### PR TITLE
Fixed device placement for Llama-2 and added tuning in bfloat16

### PR DESCRIPTION
**Pull request by Godofnothing**
Summary of changes:

1) Fixed bug that leads to device mistmatch in `finetune.py`
2) Added tuning in bfloat16 that may reduce memory usage. 
3) Removed `--temperature` argument.

Below mem usage for several options is shown.
<google-sheets-html-origin>
dtype | amp | checkpoint | Max Mem Alloc (GB) | Max Mem Res (GB)
-- | -- | -- | -- | --
float32 | N | Y | 51.61 | 57.91
float32 | Y | Y | 51.26 | 57.91
bfloat16 | N | Y | 44.66 | 62.41

### CODE REVIEW

File: src/modelutils.py [hunk 1]
```diff
@@ -28,11 +28,7 @@ def skip(*args, **kwargs):
 
 def dispatch_quantized_model(model):
     num_devices = torch.cuda.device_count()
-    device_map = {
-        "model.embed_tokens": 0,
-        "model.norm": num_devices - 1,
-        "lm_head": 0 if model.config.model_type == "cohere" else num_devices - 1,
-    }
+    device_map = {"model.embed_tokens": 0, "model.norm": num_devices - 1, "lm_head": 0}
```
Comment (justheuristic) [inline src/modelutils.py:31]: Sanity check: is this intended or did you accidentally undo parts from a previous PR?

**Review by justheuristic — APPROVED** [review]

File: finetune.py [hunk 2]
```diff
@@ -267,6 +270,13 @@ def finetune(model, train_loader, train_hiddens, args, device, val_loader=None,
         default=3,
         help="Terminate finetuning if loss doesn't improve after this number of epochs.",
     )
+    parser.add_argument(
```
Comment (justheuristic) [inline finetune.py:273]: sanity check: should we change this in README or should we keep it at default?

File: finetune.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def cache_hiddens(model, dataloader):
     return cached_hiddens
 
 
-def kl_div(student_hiddens, teacher_hiddens, temp):
```
Comment (justheuristic) [inline finetune.py:35]: Request: if temperature is not used, please remove it altogether.

Rationale: this saves memory. For models like gemma, creating another tensor of logprobs just to divide by temperature is extremely wasteful with vocabulary size 256k.

File: src/modelutils.py (hunk 1 above)
Comment (Godofnothing) [reply src/modelutils.py:31]: it is intended, for some reason, I get device mismatch without doing so

File: finetune.py (hunk 3 above)
Comment (Godofnothing) [reply finetune.py:35]: Nice catch, for Cohere it is also a big problem.

File: finetune.py (hunk 2 above)
Comment (Godofnothing) [reply finetune.py:273]: I'll update README

**Pull request merged**